compiler/phpdoc.cpp

```diff
@@ -21,115 +21,195 @@
-  {"@kphp-inline",               kphp_inline},
-  {"@kphp-flatten",              kphp_flatten},
-  {"@kphp-infer",                kphp_infer},
-  {"@kphp-required",             kphp_required},
-  {"@kphp-lib-export",           kphp_lib_export},
-  {"@kphp-sync",                 kphp_sync},
-  {"@kphp-should-not-throw",     kphp_should_not_throw},
-  {"@kphp-throws",               kphp_throws},
-  {"@type",                      var},
-  {"@var",                       var},
-  {"@return",                    returns},
-  {"@returns",                   returns},
-  {"@kphp-disable-warnings",     kphp_disable_warnings},
-  {"@kphp-extern-func-info",     kphp_extern_func_info},
-  {"@kphp-pure-function",        kphp_pure_function},
-  {"@kphp-template",             kphp_template},
-  {"@kphp-param",                kphp_param},
-  {"@kphp-return",               kphp_return},
-  {"@kphp-memcache-class",       kphp_memcache_class},
-  {"@kphp-immutable-class",      kphp_immutable_class},
-  {"@kphp-tl-class",             kphp_tl_class},
-  {"@kphp-const",                kphp_const},
-  {"@kphp-no-return",            kphp_noreturn},
-  {"@kphp-warn-unused-result",   kphp_warn_unused_result},
-  {"@kphp-warn-performance",     kphp_warn_performance},
-  {"@kphp-analyze-performance",  kphp_analyze_performance},
-  {"@kphp-serializable",         kphp_serializable},
-  {"@kphp-reserved-fields",      kphp_reserved_fields},
-  {"@kphp-serialized-field",     kphp_serialized_field},
-  {"@kphp-serialized-float32",   kphp_serialized_float32},
-  {"@kphp-profile",              kphp_profile},
-  {"@kphp-profile-allow-inline", kphp_profile_allow_inline},
-  {"@kphp-strict-types-enable",  kphp_strict_types_enable},
-  {"@kphp-color",                kphp_color},
-};
+struct KnownPhpDocTag {
+  unsigned int hash;
+  PhpDocType type;
+  const char *tag_name;
 
-vector<php_doc_tag> parse_php_doc(vk::string_view phpdoc) {
-  if (phpdoc.empty()) {
-    return {};
+  constexpr KnownPhpDocTag(const char *tag_name, PhpDocType type)
+    : hash(5381), type(type), tag_name(tag_name) {
+    for (const char *c = tag_name; *c; ++c) {
+      hash = (hash << 5) + hash + *c;
+    }
   }
+};
 
-  int line_num_of_function_declaration = stage::get_line();
+class AllDocTags {
+  static constexpr int N_TAGS = 35;
+  static const KnownPhpDocTag ALL_TAGS[N_TAGS];
 
-  vector<string> lines(1);
-  bool have_star = false;
-  for (char c : phpdoc) {
-    if (!have_star) {
-      if (c == ' ' || c == '\t') {
-        continue;
+public:
+  static PhpDocType name2type(const char *start, const char *end) __attribute__((always_inline)) {
+    unsigned int hash = 5381;
+    for (const char *c = start; c != end; ++c) {
+      hash = (hash << 5) + hash + *c;
+    }
+
+    for (const KnownPhpDocTag &tag: ALL_TAGS) {
+      if (tag.hash == hash) {
+        return tag.type;
       }
-      if (c == '*') {
-        have_star = true;
-        continue;
+    }
+    return PhpDocType::unknown;
+  }
```

Comment:
why not just std::unordered_map?
If your concern will be about significant performance difference, can we at least avoid code duplication?
What the magic constant? Can we use std::hash for this?

Reply:
At first, I've used unordered_map, but finally tested that linear search is better here, because @param/@var/@return are the move frequest tags in real code (90%), and they are placed at the top of `ALL_TAGS`.
A "magic number" is from http://www.cse.yorku.ca/~oz/hash.html. It's very simple but good enough for short strings (like names). 
About code duplication — personally I don't see any problem in duplicating 3 lines kept together and never changed (even if occasionally changed, lots of tests would fail). The first should be constexpr, the latter should not. How would you advice to remove duplication?

Reply:
Ok, got it about performance. 
About hash, i've found `string_hash_5381()` in our code with same logic. If you make it constexpr, noexcept, accepted std::string_view then it will be possible to use it in both places. E.g. constructor will looks like:
```
  constexpr KnownPhpDocTag(const char *tag_name, PhpDocType type) noexcept
    : hash(string_hash_5381(tag_name))
    , type(type)
    , tag_name(tag_name) {}
```